content/source/docs/plugin/framework/validation.html.md

```diff
@@ -0,0 +1,261 @@
+        )
+
+        return
+    }
+}
+```
+
+Optionally and depending on the complexity, it may be desirable to also create a helper function to instantiate the validator. For example:
+
+```go
+func stringLengthBetween(minLength int, maxLength int) stringLengthBetweenValidator {
+    return stringLengthBetweenValidator{
+        Max: maxLength,
+        Min: minLength,
+    }
+}
+```
+
+## Type Validation
+
+Providers that contain common attribute values with consistent validation rules may wish to create a custom type to simplify schemas. When validation is implemented on a type, it is not necessary to declare the same validation on the attribute, although additional validations can be supplied in that manner. For example:
+
+```go
+// Typically within the tfsdk.Schema returned by GetSchema() for a provider,
+// resource, or data source.
+tfsdk.Attribute{
+    // ... other Attribute configuration ...
+
+    // This is an example type which implements its own validation
+    Type: computeInstanceIdentifierType,
+
+    // This is optional, example validation that is checked in addition
+    // to any validation performed by the type
+    Validators: []AttributeValidators{
+        stringLengthBetween(10, 256),
+    },
+}
+```
+
+### Defining Type Validation
+
+To support validation within a type, the [`attr.TypeWithValidate` interface](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-framework/attr#TypeWithValidate) must be satisfied. For example:
+
+```go
+// Other methods to implement the attr.Type interface are omitted for brevity
+type computeInstanceIdentifierType struct {}
+
+func (t computeInstanceIdentifierType) Validate(ctx context.Context, tfValue tftypes.Value, path *tftypes.AttributePath) diag.Diagnostics {
+    var diags diag.Diagnostics
+
+    if !tfValue.Type().Equal(tftypes.String) {
+        diags.AddAttributeError(
+            path,
+            "Compute Instance Type Validation Error",
+            fmt.Sprintf("Expected String value, received %T with value: %v", in, in),
+        )
+        return diags
+    }
+
+    if !tfValue.IsKnown() || tfValue.IsNull() {
+        return diags
+    }
+
+    var value string
+    err := tfValue.As(&value)
+
+    if err != nil {
+        diags.AddAttributeError(
+            path,
+            "Compute Instance Type Validation Error",
+            fmt.Sprintf("Cannot convert value to string: %s", err),
+        )
+        return diags
+    }
+
+    if !strings.HasPrefix(value, "instance-") {
+        diags.AddAttributeError(
+            path,
+            "Compute Instance Type Validation Error",
+            fmt.Sprintf("Missing `instance-` prefix, got: %s", value),
+        )
+        return diags
+    }
+}
+```
+
+## Schema Validation
+
+Provider, resource, and data source schemas also support validation across all attributes. This is helpful when checking values in multiple attributes, such as only accepting certain values in one attribute when another is a specific value.
```

Comment:
Could we make the example after "such as" more specific/a real-world scenario?

Reply:
Maybe this will be slightly better (it is a little hard because these tend to be very-circumstance specific):

```suggestion
Provider, resource, and data source schemas also support validation across all attributes. This is helpful when checking values in multiple attributes, such as ensuring the values are compatible with each other.
```